pkg/oc/graph/imagegraph/nodes/nodes_test.go

```diff
@@ -0,0 +1,106 @@
+package nodes
+
+import (
+	"testing"
+
+	"github.com/openshift/origin/pkg/oc/admin/prune/imageprune/testutil"
+	"github.com/openshift/origin/pkg/oc/graph/genericgraph"
+)
+
+const registryHost = "registry.io"
+
+func TestEnsureImageNode(t *testing.T) {
+	g := genericgraph.New()
+
+	img1 := testutil.Image("sha256:0000000000000000000000000000000000000000000000000000000000000001", registryHost+"/foo/bar@sha256:0000000000000000000000000000000000000000000000000000000000000001")
+	img2 := testutil.Image("sha256:0000000000000000000000000000000000000000000000000000000000000002", registryHost+"/foo/bar@sha256:0000000000000000000000000000000000000000000000000000000000000002")
+	img3 := testutil.Image("sha356:0000000000000000000000000000000000000000000000000000000000000003", registryHost+"/foo/bar@sha356:0000000000000000000000000000000000000000000000000000000000000003")
+
+	node := FindImage(g, img1.Name)
+	if node != nil {
+		t.Fatalf("got unexpected image")
+	}
+	n1 := EnsureImageNode(g, &img1)
+	if n1 == nil {
+		t.Fatalf("failed to create node")
+	}
+	node = FindImage(g, img1.Name)
+	if node == nil {
+		t.Fatalf("expected node, not nil")
+	}
+	if node.Image != &img1 {
+		t.Fatalf("got unexpected image: %s != %s", node.Image.Name, img1.Name)
+	}
+
+	n2 := EnsureImageNode(g, &img2)
+	if n2 == nil {
+		t.Fatalf("expected node, not nil")
+	}
+	if len(g.Nodes()) != 2 {
+		t.Fatalf("unexpected number of nodes: %d != %d", len(g.Nodes()), 2)
+	}
+
+	g.RemoveNode(n1)
+
+	// removing node deletes its edges but the image is still known
+	node = FindImage(g, img1.Name)
+	if node == nil {
+		t.Fatalf("removed node can still be found")
+	}
+	// but it's no longer iterable
+	if len(g.Nodes()) != 1 {
+		t.Fatalf("unexpected number of nodes: %d != %d", len(g.Nodes()), 1)
+	}
+
+	// add the image1 again
+	n1 = EnsureImageNode(g, &img1)
+	if n1 == nil {
+		t.Fatalf("failed to create node")
+	}
+	node = FindImage(g, img1.Name)
+	if node == nil {
+		t.Fatalf("expected node, not nil")
+	}
+	if node.Image != &img1 {
+		t.Fatalf("got unexpected image: %s != %s", node.Image.Name, img1.Name)
+	}
+	/**
+	 * NOTE: it looks like removed node cannot be re-added
+	if len(g.Nodes()) != 2 {
```

Comment:
While adding this test I found another limitation of the graph algorithm. The consequence is that image or image stream that is re-added after being pruned will be ignored.

I'd say that the odds for the same image to be added while being pruned is pretty low but it's still unfortunate.